apache/superset
# feat: generate consistent QueryObject whether GenericAxis is enabled or disabled

### SUMMARY

This PR intends to generate a consistent QueryObject whether GenericAxis FF is enabled or disabled. e.g.: the charts wouldn't change on the Dashboard although Axis FF from `enabled` to `disabled` and vice versa.

1. Make sure that the charts on the Dashboards are consistent whether FF is enabled or disabled, so the FF GenericAxis should be safety enabled by default (will do in separated PR).
2. Add unit test in Echart timeseries charts.
3. Add unit test in Echart MixedTimeseries charts.
4. Move `getXAxis` and `isXAxisSet` into `superset-ui/core`

There are 2 kinds of `queryObject` patterns after this PR:

#### The QueryObjec will be like below when the `x_axis` is not in formData whether `GenericAxis` is enabled or disabled
```
{
        granularity: 'time_column',
        extras: { time_grain_sqla: 'P1Y', having: '', where: '' },
        is_timeseries: true,
        post_processing: [
          {
            operation: 'pivot',
            options: {
              aggregates: { 'count(*)': { operator: 'mean' } },
              columns: ['col1'],
              drop_missing_columns: true,
              index: ['__timestamp'],
            },
          },
          { operation: 'flatten' },
        ],
      }
     ....
     ....
}
```

#### The QueryObjec will be like below when the `x_axis` is in formData whether `GenericAxis` is enabled or disabled.
The `granularity`(axis) and `time_grain` will be converted to `BASE_AXIS`, then remove `time_grain_sqla` and `is_timeseries`

```
{
        extras: { having: '', where: '' },
        columns: [
          {
            columnType: 'BASE_AXIS',
            expressionType: 'SQL',
            label: 'time_column',
            sqlExpression: 'time_column',
            timeGrain: 'P1Y',
          },
          'col1',
        ],
        post_processing: [
          {
            operation: 'pivot',
            options: {
              aggregates: { 'count(*)': { operator: 'mean' } },
              columns: ['col1'],
              drop_missing_columns: true,
              index: ['time_column'],
            },
          },
          { operation: 'flatten' },
        ],
      }
     ....
     ....
```

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


### TESTING INSTRUCTIONS

#### Test Case 1
1. Turn off GenericAxis
2. Make an Echart time-series chart and a Mixed time-series, save to a new Dashboard.
3. Turn on GenericAxis
4. Refresh Dashboard
5. Charts should be as before

#### Test Case 2
1. Turn on GenericAxis
2. Make an Echart time-series chart and a Mixed time-series, save to a new Dashboard.
3. Turn off GenericAxis
4. Refresh Dashboard
5. Charts should be as before



### ADDITIONAL INFORMATION


- [ ] Introduces a new feature or API
